src/main/java/org/jvnet/hudson/test/junit/jupiter/JenkinsRuleResolver.java

```diff
@@ -0,0 +1,43 @@
+package org.jvnet.hudson.test.junit.jupiter;
+
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.jvnet.hudson.test.JenkinsRule;
+
+/**
+ * JUnit 5 extension providing {@link JenkinsRule} integration.
+ */
+public class JenkinsRuleResolver implements ParameterResolver, AfterEachCallback {
```

Comment:
Both rule annotations are superseded by [`@ExtendWith` / `@RegisterExtension`](https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4-tips).

According to the [docs](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/extension/ExtendWith.html), `@ExtendWith` can be used on test class and method level.

Reply:
Best to test that explicitly.

Reply:
I have done a quick test: `@JenkinsRule` (= `@ExentdsWith`) can either be used on a class or specific method. If placed on the former, all tests within can access it; if on the latter, only the annotated test method can.

Or do you mean additional unit tests to verify this behaviour?

Reply:
> can either be used on a class or specific method

That is something very different. (See the occasionally used `@WithoutJenkins` in 4 tests, though `JenkinsSessionRule` is the cleaner solution since you can explicitly control whether and when Jenkins starts and shuts down.)

I (and perhaps @uhafner also) was referring to the difference between

```java
public class XTest {
  @Rule public JenkinsRule r = new JenkinsRule();
  @Test public one() {
    WorkflowJob p = r.createProject(WorkflowJob.class, "p");
    // …
  }
  @Test public two() {
    WorkflowJob p = r.createProject(WorkflowJob.class, "p");
    // …
  }
}
```

and

```java
public class XTest {
  @ClassRule public static JenkinsRule r = new JenkinsRule();
  @Test public one() {
    WorkflowJob p = r.createProject(WorkflowJob.class, "p1");
    // …
  }
  @Test public two() {
    WorkflowJob p = r.createProject(WorkflowJob.class, "p2");
    // …
  }
}
```

`@ClassRule` is a bit faster since Jenkins only starts up once for the whole suite. (The effect is not as big as you might think, since most of the overhead is class loading, which applies only to the first test run—unlike `RealJenkinsRule`, Jenkins core & plugin classes are loaded in the app class loader.) But you must take care that tests do not interfere with one another. `@Rule` is safer since every test gets its own copy of Jenkins (incl. `$JENKINS_HOME`).

> do you mean additional unit tests to verify this behaviour

I did mean an automated test yes.

Reply:
Thanks for your detailed explanation. Now, only the behaviour of `@Rule` is supported – each test gets it's own `JenkinsRule` object.

Reply:
Fine for now but should be documented explicitly.

Reply:
Yes, I meant the same thing. Thanks for providing the example! My integration tests are lasting quite some time so I tried to optimize the runtime. And using a `ClassRule` helps to reduce the runtime by around 20%. 

But anyway, I think this is a good first iteration, where the most important use case is working now. Thanks for adding it! I think this is a good basis and can be extended later with additional PRs.

Reply:
Documentation updated (as suggested by @uhafner below) and there's a test for method level annotation too now.